### PR TITLE
Move `GetOuputString()` and `GetFormatStyleString()` to PSHostUserInterface as public API

### LINKED ISSUE #14387: Expose the `Utils.GetOutputString` as a public method for C# developers to work with `OutputRendering`
This issue is used to track the work to address the review comment https://github.com/PowerShell/PowerShell/pull/13758#discussion_r540412372. 

> This API may need to be public, so that other host implementations can strip off ANSI sequences the same way PowerShell does. All existing powershell hosts will need it when moving to PS 7.2, otherwise, their users will find `$PSStyle.OutputRendering` doesn't take effect. But I'm fine to expose the API in a separate PR. Let's open an issue to track it.
>
> There is a need, for sure! I have developed my own versions of this as there has been no API.
I even think we should consider an API that doesn't just expose the implementation details but to think about what devs would like to do with it.
Otherwise, everyone will have to invent that wheel over and over again.
>
>Another opportunity if you use methods is that you can determine if ANSI-output isn't enabled and then just bypass the decorations.

I think maybe the following 2 APIs need to be made public
- `Utils.GetOutputString` (overloads may be needed)
  - The `isOutputRedirected` parameter needs to be revisited, at least its name. For a different host implementation, `Console.IsOutputRedirected == true` doesn't necessarily means it wants plain text. Think about the host of PS jupyter kernel in .NET Interactive, the process's output/error streams are always redirected to send to the frontend -- jupyter notebook, which is capable of handling ANSI sequences. So, in this case, even thought the std out/err are redirected, the ANSI sequences shouldn't be stripped off.
- `Utils.GetFormatStyleString`
  - This method currently directly check on `Console.IsErrorRedirected` and `Console.IsOutputRedirected`. I think we shouldn't directly check on those in PowerShell engine, but should instead have separate parameters for the caller to provide that information. The engine should not assume there is a console available, and it's up to the host implementation to pass in whether there is error/output redirection when calling this method.
- ~`Utils.ShouldOutputPlainText` (not sure if this one needs to be public too, but should evaluate it)~ probably not needed to be public.

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
# PR Summary

Fix #14387

Follow up on #14397. These two APIs were previously internal and used by the ConsoleHost.  However, this means that every host would need to have similar code to handle ANSI rendering correctly.  Discussion with @daxian-dbw and @rjmholt was to move them to `PSHostUserInterface` so they can be used by the dotnet-interactive Jupyter kernel and the vscode-powershell extension.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: Reference docs are generated from the source code XML comments
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs [hunk 1]
```diff
@@ -232,6 +232,159 @@ public virtual void WriteLine(ConsoleColor foregroundColor, ConsoleColor backgro
...
+        /// <param name="supportsVirtualTerminal">
+        /// True if the host supports virtual terminal.
+        /// </param>
+        /// <param name="isOutputRedirected">
+        /// True if the output is redirected.
+        /// </param>
+        /// <returns>
+        /// The formatted text.
+        /// </returns>
+        public static string GetOutputString(string text, bool isHost, bool? supportsVirtualTerminal = null, bool isOutputRedirected = false)
```
Comment (rjmholt) [inline src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs:365]: [Public APIs shouldn't have default parameters](https://rules.sonarsource.com/csharp/RSPEC-2360)

**Comment by SteveL-MSFT** [comment]
I'm going to make a change, for public hosts, some of the parameters don't make any sense.  I'll have a simpler public version and keep the current one as internal.

File: src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs [hunk 2]
```diff
@@ -232,6 +232,162 @@ public virtual void WriteLine(ConsoleColor foregroundColor, ConsoleColor backgro
...
+        /// <param name="formatStyle">
+        /// The format style to get the escape sequence for.
+        /// </param>
+        /// <param name="isOutputRedirected">
+        /// True if the output is redirected.
+        /// </param>
+        /// <returns>
+        /// The ANSI escape sequence for the given format style.
+        /// </returns>
+
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs:313]:
```suggestion
```

File: src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs [hunk 3]
```diff
@@ -232,6 +232,162 @@ public virtual void WriteLine(ConsoleColor foregroundColor, ConsoleColor backgro
...
+        public static string GetFormatStyleString(FormatStyle formatStyle, bool isOutputRedirected)
+        {
+            // redirected console gets plaintext output to preserve existing behavior
+            if (!InternalTestHooks.BypassOutputRedirectionCheck &&
+                ((PSStyle.Instance.OutputRendering == OutputRendering.PlainText) ||
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/hostifaces/MshHostUserInterface.cs:318]:
```suggestion
                (PSStyle.Instance.OutputRendering == OutputRendering.PlainText ||
```

**Review by daxian-dbw — APPROVED** [review]

**Pull request merged**